src/rules.jl

```diff
@@ -184,13 +184,25 @@ DNERule(args...) = DNE()
 #####
 
 """
-TODO
+    WirtingerRule([𝒟::Type, ]P::AbstractRule, C::AbstractRule)
+Construct a `WirtingerRule` object, which is an `AbstractRule` that consists of
+an `AbstractRule` for both the primal derivative ``∂/∂x`` and the conjugate
+derivative ``∂/∂x̅``. If the domain `𝒟` is specified, return a `Rule` evaluating
+to `P(Δ) + C(Δ)` if `𝒟 <: Real`, otherwise return `WirtingerRule(P, C)`.
 """
 struct WirtingerRule{P<:AbstractRule,C<:AbstractRule} <: AbstractRule
     primal::P
     conjugate::C
 end
 
+function WirtingerRule(𝒟::Type, primal::AbstractRule, conjugate::AbstractRule)
+    if 𝒟 <: Real || eltype(𝒟) <: Real
```

Comment:
Can't trust `eltype` in general.
Might be a `Vector{Any}` that only contains real numbers

Reply:
But then I don't think there's a type stable way to test for this case. This is only an optimization anyway, so I'm not to concerned, as returning `Wirtinger` will always be a safe fallback.

Reply:
If it is always safe left's do that. I expect the constructor `WirtingerRule` to return a `WirtingerRule`.
(You can actually break `Dict` of generator expressions if you violate that. Though this is arguably a bug in Base)

Reply:
If it is always safe left's do that. I expect the constructor `WirtingerRule` to return a `WirtingerRule`.
(You can actually break `Dict` of generator expressions if you violate that. Though this is arguably a bug in Base)

Reply:
Yes, I agree that returning a different type from a constructor feels a bit weird, but then again, if you're explicitly specifying the domain type, you're willingly opting into this, as you can just leave it off. Do you think that this should be a separate function, maybe just lowercased?

Reply:
If your explictly doing it, can't you just explictly use a normal `Rule` ?

Reply:
You could of course do this check yourself, but for `abs2` for example, this saves you from having to write special rules for a real input, so that you don't unnecessarily propagate `Wirtinger` unless you have to.

Reply:
@jrevels I don't know anyone who really understands the Wirtinger stuff well enough to review this further. Can you take a look?